include/unifex/std_concepts.hpp

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/detail/concept_macros.hpp>
+#include <unifex/swap.hpp>
+
+#include <functional>
+#include <type_traits>
+
+#if defined(_MSC_VER) && !defined(__clang__)
+  #define UNIFEX_DIAGNOSTIC_PUSH __pragma(warning(push))
+  #define UNIFEX_DIAGNOSTIC_POP __pragma(warning(pop))
+  #define UNIFEX_DIAGNOSTIC_IGNORE_INIT_LIST_LIFETIME
+  #define UNIFEX_DIAGNOSTIC_IGNORE_FLOAT_EQUAL
+  #define UNIFEX_DIAGNOSTIC_IGNORE_CPP2A_COMPAT
+#else // ^^^ defined(_MSC_VER) ^^^ / vvv !defined(_MSC_VER) vvv
+  #if defined(__GNUC__) || defined(__clang__)
+    #define UNIFEX_PRAGMA(X) _Pragma(#X)
+    #define UNIFEX_DIAGNOSTIC_PUSH UNIFEX_PRAGMA(GCC diagnostic push)
+    #define UNIFEX_DIAGNOSTIC_POP UNIFEX_PRAGMA(GCC diagnostic pop)
+    #define UNIFEX_DIAGNOSTIC_IGNORE_PRAGMAS \
+      UNIFEX_PRAGMA(GCC diagnostic ignored "-Wpragmas")
+    #define UNIFEX_DIAGNOSTIC_IGNORE(X) \
+      UNIFEX_DIAGNOSTIC_IGNORE_PRAGMAS \
+      UNIFEX_PRAGMA(GCC diagnostic ignored "-Wunknown-pragmas") \
+      UNIFEX_PRAGMA(GCC diagnostic ignored X)
+    #define UNIFEX_DIAGNOSTIC_IGNORE_INIT_LIST_LIFETIME \
+      UNIFEX_DIAGNOSTIC_IGNORE("-Wunknown-warning-option") \
+      UNIFEX_DIAGNOSTIC_IGNORE("-Winit-list-lifetime")
+    #define UNIFEX_DIAGNOSTIC_IGNORE_FLOAT_EQUAL UNIFEX_DIAGNOSTIC_IGNORE("-Wfloat-equal")
+    #define UNIFEX_DIAGNOSTIC_IGNORE_CPP2A_COMPAT UNIFEX_DIAGNOSTIC_IGNORE("-Wc++2a-compat")
+  #else
+    #define UNIFEX_DIAGNOSTIC_PUSH
+    #define UNIFEX_DIAGNOSTIC_POP
+    #define UNIFEX_DIAGNOSTIC_IGNORE_INIT_LIST_LIFETIME
+    #define UNIFEX_DIAGNOSTIC_IGNORE_FLOAT_EQUAL
+    #define UNIFEX_DIAGNOSTIC_IGNORE_CPP2A_COMPAT
+  #endif
+#endif // MSVC/Generic configuration switch
```

Comment:
This doesn't seem particularly specific to std_concepts being defined here.
Should it be moved to `config.h.in`?